backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/webservice/schema/CardMutationService.kt

```diff
@@ -115,6 +124,88 @@ class CardMutationService {
         }
     }
 
+    @GraphQLDescription("Creates a new digital entitlementcard from self-service portal")
+    fun createCardFromSelfService(
+        dfe: DataFetchingEnvironment,
+        project: String,
+        encodedCardInfo: String,
+        generateStaticCode: Boolean
+    ): CardCreationResultModel {
+        // TODO disable for production until the Koblenz project is ready to go live
+        val context = dfe.getContext<GraphQLContext>()
+        if (context.backendConfiguration.production) {
+            throw ForbiddenException()
+        }
+        if (project != KOBLENZ_PASS_PROJECT) {
+            throw NotKoblenzProjectException()
+        }
+
+        val cardInfo = parseEncodedCardInfo(encodedCardInfo)
+        val regionId = cardInfo.extensions.extensionRegion.regionId
+        transaction { RegionsRepository.findByIdInProject(project, regionId) } ?: throw RegionNotFoundException()
+
+        val user = KoblenzUser(
+            cardInfo.fullName,
+            cardInfo.extensions.extensionBirthday.birthday,
+            cardInfo.extensions.extensionKoblenzReferenceNumber.referenceNumber
+        )
+
+        val userHash = Argon2IdHasher.hashKoblenzUserData(user)
+
+        val userEntitlements = transaction { UserEntitlementsRepository.findUserEntitlements(userHash.toByteArray(), regionId) }
+        if (userEntitlements == null || userEntitlements.revoked || userEntitlements.endDate.isBefore(LocalDate.now())) {
+            throw InvalidUserEntitlementsException()
+        }
+
+        val cardInfoWithExpDayAndStartDay = enrichCardInfoWithExpDayAndStartDay(
+            cardInfo,
+            userEntitlements.endDate.toEpochDay().toInt(),
+            userEntitlements.startDate.toEpochDay().toInt()
+        )
+
+        val activationCode = CardCreationResultModel(
+            createDynamicActivationCode(cardInfoWithExpDayAndStartDay, userId = null),
+            if (generateStaticCode) createStaticVerificationCode(cardInfoWithExpDayAndStartDay, userId = null) else null
+        )
+
+        val projectConfig = context.backendConfiguration.projects.find { it.id == project } ?: throw ProjectNotFoundException(project)
+
+        Matomo.trackCreateCards(
+            projectConfig,
+            context.request,
+            dfe.field.name,
+            regionId,
+            numberOfDynamicCards = 1,
+            numberOfStaticCards = if (generateStaticCode) 1 else 0
+        )
+
+        return activationCode
+    }
+
+    private fun enrichCardInfoWithExpDayAndStartDay(cardInfo: Card.CardInfo, expirationDay: Int, startDay: Int): Card.CardInfo {
+        return cardInfo.toBuilder()
```

Comment:
Maybe there is a less ugly way to update card info to set dates from the user entitlements table?